### PR TITLE
Introduce a join point for SdkLocator and VSTemplateLocator

### LINKED ISSUE #4336: Determine how to build the SdkLocator and VSTemplateLocator in the VMR
The SdkResolver and the VSTemplateLocator projects in dotnet/installer both consume all Windows architectures of the hostfxr binary to provide a single package to ship to VS that can locate the .NET SDKs and templates for representation within VS. In the VMR individual jobs, we only have the current RID available when building.

### LINKED ISSUE #4633: Create first pass 2 stage and job
Create a pass 2 stage where the first join point jobs can be implemented.

### PR DESCRIPTION
**Pull request by directhex**
For full builds (i.e. all lanes, not the reduced CI-only set), this switches the Windows x64 build to second pass, so it can ingest packages from the Windows x86 and Windows ARM64 builds

Closes: https://github.com/dotnet/source-build/issues/4336
Closes: https://github.com/dotnet/source-build/issues/4633

### CODE REVIEW

File: src/SourceBuild/content/repo-projects/sdk.proj [hunk 1]
```diff
@@ -25,6 +25,9 @@
     <BuildArgs>$(BuildArgs) /p:Architecture=$(TargetArchitecture)</BuildArgs>
     <BuildArgs>$(BuildArgs) /p:DOTNET_INSTALL_DIR=$(DotNetRoot)</BuildArgs>
 
+    <!-- Pass through DotNetBuildPass due to conditional behaviour for this project on join points -->
+    <BuildArgs Condition="'$(DotNetBuildPass)' != ''">$(BuildArgs) /p:DotNetBuildPass=$(DotNetBuildPass)</BuildArgs>
```
Comment (mmitche) [inline src/SourceBuild/content/repo-projects/sdk.proj:29]: I think this can be hoisted out to Directory.Build.props.

File: eng/DotNetBuild.props [hunk 2]
```diff
@@ -17,6 +17,7 @@
       <InnerBuildArgs>$(InnerBuildArgs) /p:IncludeAdditionalSharedFrameworks=false</InnerBuildArgs>
       <InnerBuildArgs Condition="'$(SourceBuildUseMonoRuntime)' == 'true'">$(InnerBuildArgs) /p:DISABLE_CROSSGEN=true</InnerBuildArgs>
       <InnerBuildArgs Condition="'$(PgoInstrument)' == 'true'">$(InnerBuildArgs) /p:PgoInstrument=true</InnerBuildArgs>
+      <InnerBuildArgs Condition="'$(DotNetBuildPass)' != ''">$(InnerBuildArgs) /p:DotNetBuildPass=$(DotNetBuildPass)</InnerBuildArgs>
```
Comment (mmitche) [inline eng/DotNetBuild.props:20]: This should be forwarded in the inner arcade command buildup: https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets#L67

You can keep it here for now though.

File: src/Installer/redist-installer/projects/SdkResolver.csproj [hunk 3]
```diff
@@ -7,11 +7,7 @@
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
     <GenerateRuntimeConfigurationFiles>false</GenerateRuntimeConfigurationFiles>
     <ExcludeFromSourceOnlyBuild>true</ExcludeFromSourceOnlyBuild>
-    <!--
-      References assets from multiple architectures and can't be built inside the VMR until we have a join point job
-      that can pull assets from multiple legs. https://github.com/dotnet/source-build/issues/4336
-    -->
-    <ExcludeFromDotNetBuild>true</ExcludeFromDotNetBuild>
+    <ExcludeFromDotNetBuild Condition="'$(DotNetBuildPass)' == ''">true</ExcludeFromDotNetBuild>
```
Comment (mmitche) [inline src/Installer/redist-installer/projects/SdkResolver.csproj:10]: Do you think there is likely to be a pass 3 (that isn't the merge pass)? If so, then I think we should be explicit about this being pass 2.

**Comment by mmitche** [comment]
Does this exclude all the other sdk projects in pass 2? @ViktorHofer built a bit of infra to do this: https://github.com/dotnet/arcade/blob/f88082a3a89505296ec57f1a5f94fdc407fe1e58/Documentation/UnifiedBuild/Unified-Build-Join-Point.md?plain=1#L109

I think you may just need to include those two csprojs in the pass 2 itemgroup.

File: src/Installer/redist-installer/projects/SdkResolver.csproj (hunk 3 above)
Comment (directhex) [reply src/Installer/redist-installer/projects/SdkResolver.csproj:10]: I think it depends. If we think of pass 2 as "a pass that needs artifacts from pass 1 on other verticals" and pass 3 as "a pass that needs artifacts from pass 2 on other verticals", there _could_ be scenarios where that's needed, especially around workloads.

Do we _want_ to constrain this to only ever work in pass 2, or simply "not in pass 1, which by definition only has one vertical and therefore won't include the required packages"

**Review by mmitche — APPROVED** [review]

File: src/Installer/redist-installer/projects/SdkResolver.csproj (hunk 3 above)
Comment (ViktorHofer) [reply src/Installer/redist-installer/projects/SdkResolver.csproj:10]: Yes, we should be explicit and only produce and upload artifacts that are needed in that build pass. I will submit a PR to fine tune this.

File: eng/DotNetBuild.props (hunk 2 above)
Comment (ViktorHofer) [reply eng/DotNetBuild.props:20]: This property shouldn't be needed at all. Global properties are automatically passed from the outer to the inner build.

**Comment by ViktorHofer** [comment]
Submitted https://github.com/dotnet/sdk/pull/44153 as a follow-up.

**Pull request merged**